framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,635 @@
+        printWriter.print("(");
+
+        StringJoiner parameters = new StringJoiner(", ");
+
+        if (!aMethod.receiver.type.tlAnnotationsHere.isEmpty()) {
+            // Only output the receiver if it has an annotation.
+            parameters.add(
+                    formatParameter(
+                            AFieldWrapper.createReceiverParameter(aMethod.receiver, basename),
+                            "this",
+                            basename));
+        }
+        for (Integer index : aMethodWrapper.getParameters().keySet()) {
+            AFieldWrapper param = aMethodWrapper.getParameters().get(index);
+            parameters.add(formatParameter(param, param.getParameterName(), basename));
+        }
+        printWriter.print(parameters.toString());
+        printWriter.println(");");
+        printWriter.println();
+    }
+
+    /**
+     * The implementation of {@link #write(ASceneWrapper, Writer)} and {@link #write(ASceneWrapper,
+     * String)}. Prints imports, classes, method signatures, and fields in stub file format, all
+     * with appropriate annotations.
+     *
+     * @param scene the scene to write
+     * @param printWriter where to write the scene to
+     */
+    private static void writeImpl(ASceneWrapper scene, PrintWriter printWriter) {
+        // Write out all imports
+        ImportDefWriter importDefWriter;
+        try {
+            importDefWriter = new ImportDefWriter(scene, printWriter);
+        } catch (DefException e) {
+            throw new BugInCF(e);
+        }
+        importDefWriter.visit();
+        printWriter.println();
+
+        // For each class
+        class_loop:
+        for (Map.Entry<@BinaryName String, AClassWrapper> classEntry :
+                scene.getClasses().entrySet()) {
+            printClass(classEntry, printWriter);
+        }
+        printWriter.flush();
+    }
+
+    /**
+     * Print the class body, or nothing if this is an anonymous inner class
+     *
+     * @param classEntry the class to print, as a Map entry. The key is the class name in binary
+     *     form. The value is the AClassWrapper object representing the class.
+     * @param printWriter the writer on which to print
+     */
+    private static void printClass(
+            Map.Entry<@BinaryName String, AClassWrapper> classEntry, PrintWriter printWriter) {
+
+        String classname = classEntry.getKey();
+        String basename = basenamePart(classname);
+
+        if ("package-info".equals(basename) || "module-info".equals(basename)) {
+            return;
+        }
+
+        // Do not attempt to print stubs for anonymous inner classes, because the stub parser
+        // cannot read them. (An anonymous inner class has a basename like Outer$1, so this
+        // check ensures that the binary name's final segment after its last $ is not only
+        // composed of digits.)
+        String innermostClassname = basename;
+        if (innermostClassname.contains("$")) {
```

Comment:
This is clumsy and hard to follow.  If you are going to use a regular expression, use just one:
`\$\d+$`
Although that is equivalent to what you have written, it doesn't omit named inner classes of anonymous classes.  Instead, use
`\$\d+(\$|$)`